### PR TITLE
CUDA: Explicitly specify objmode + looplifting for jit functions in cuda.random

### LINKED ISSUE #4809: Warnings executing numba random number generator example with simulator enabled
Numba version: 0.46.0

Executing the example posted in "http://numba.pydata.org/numba-doc/latest/cuda/random.html#example" with:

```
python random_testing.py
```
Works well, but executing it with CUDA simulator enabled:

```
NUMBA_ENABLE_CUDASIM=1 python random_testing.py
```

I get NumbaDeprecationWarning and RuntimeWarning:
```
/home/jetxeberria/documents/programs/anaconda/anaconda3/envs/testing3/lib/python3.7/site-packages/numba/cuda/random.py:135: NumbaWarning: 
Compilation is falling back to object mode WITH looplifting enabled because Function "xoroshiro128p_uniform_float32" failed type inference due to: non-precise type pyobject
[1] During: typing of argument at /home/jetxeberria/documents/programs/anaconda/anaconda3/envs/testing3/lib/python3.7/site-packages/numba/cuda/random.py (145)

File "../../../../../../../../../../../../home/jetxeberria/documents/programs/anaconda/anaconda3/envs/testing3/lib/python3.7/site-packages/numba/cuda/random.py", line 145:
def xoroshiro128p_uniform_float32(states, index):
    <source elided>
    '''
    index = int64(index)
    ^

  @jit
/home/jetxeberria/documents/programs/anaconda/anaconda3/envs/testing3/lib/python3.7/site-packages/numba/object_mode_passes.py:178: NumbaWarning: Function "xoroshiro128p_uniform_float32" was compiled in object mode without forceobj=True.

File "../../../../../../../../../../../../home/jetxeberria/documents/programs/anaconda/anaconda3/envs/testing3/lib/python3.7/site-packages/numba/cuda/random.py", line 136:
@jit
def xoroshiro128p_uniform_float32(states, index):
^

  state.func_ir.loc))
/home/jetxeberria/documents/programs/anaconda/anaconda3/envs/testing3/lib/python3.7/site-packages/numba/object_mode_passes.py:187: NumbaDeprecationWarning: 
Fall-back from the nopython compilation path to the object mode compilation path has been detected, this is deprecated behaviour.

For more information visit http://numba.pydata.org/numba-doc/latest/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit

File "../../../../../../../../../../../../home/jetxeberria/documents/programs/anaconda/anaconda3/envs/testing3/lib/python3.7/site-packages/numba/cuda/random.py", line 136:
@jit
def xoroshiro128p_uniform_float32(states, index):
^

  warnings.warn(errors.NumbaDeprecationWarning(msg, state.func_ir.loc))
/home/jetxeberria/documents/programs/anaconda/anaconda3/envs/testing3/lib/python3.7/site-packages/numba/cuda/random.py:71: NumbaWarning: 
Compilation is falling back to object mode WITH looplifting enabled because Function "xoroshiro128p_next" failed type inference due to: non-precise type pyobject
[1] During: typing of argument at /home/jetxeberria/documents/programs/anaconda/anaconda3/envs/testing3/lib/python3.7/site-packages/numba/cuda/random.py (81)

File "../../../../../../../../../../../../home/jetxeberria/documents/programs/anaconda/anaconda3/envs/testing3/lib/python3.7/site-packages/numba/cuda/random.py", line 81:
def xoroshiro128p_next(states, index):
    <source elided>
    '''
    index = int64(index)
    ^

  @jit
/home/jetxeberria/documents/programs/anaconda/anaconda3/envs/testing3/lib/python3.7/site-packages/numba/object_mode_passes.py:178: NumbaWarning: Function "xoroshiro128p_next" was compiled in object mode without forceobj=True.

File "../../../../../../../../../../../../home/jetxeberria/documents/programs/anaconda/anaconda3/envs/testing3/lib/python3.7/site-packages/numba/cuda/random.py", line 72:
@jit
def xoroshiro128p_next(states, index):
^

  state.func_ir.loc))
/home/jetxeberria/documents/programs/anaconda/anaconda3/envs/testing3/lib/python3.7/site-packages/numba/object_mode_passes.py:187: NumbaDeprecationWarning: 
Fall-back from the nopython compilation path to the object mode compilation path has been detected, this is deprecated behaviour.

For more information visit http://numba.pydata.org/numba-doc/latest/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit

File "../../../../../../../../../../../../home/jetxeberria/documents/programs/anaconda/anaconda3/envs/testing3/lib/python3.7/site-packages/numba/cuda/random.py", line 72:
@jit
def xoroshiro128p_next(states, index):
^

  warnings.warn(errors.NumbaDeprecationWarning(msg, state.func_ir.loc))
random_testing.py:16: RuntimeWarning: overflow encountered in ulong_scalars
  x = xoroshiro128p_uniform_float32(rng_states, thread_id)
random_testing.py:17: RuntimeWarning: overflow encountered in ulong_scalars
  y = xoroshiro128p_uniform_float32(rng_states, thread_id)
random_testing.py:16: RuntimeWarning: overflow encountered in ulong_scalars
  x = xoroshiro128p_uniform_float32(rng_states, thread_id)
pi: 3.1367188
```

### PR DESCRIPTION
**Pull request by gmarkall**
Using the `@jit` decorator alone results in warnings being emitted when using the `cuda.random` functions with the simulator enabled, because Fake CUDA Arrays are passed to jitted functions. Since these can't be typed or unboxed, a fallback to object mode with looplifting is needed.

To avoid the warning, we explicitly specify that object mode with looplifting should be used when the simulator is enabled.

Fixes #4809.

### CODE REVIEW

File: numba/cuda/random.py [hunk 1]
```diff
@@ -33,8 +33,16 @@
                                align=True)
 xoroshiro128p_type = from_dtype(xoroshiro128p_dtype)
 
+# When cudasim is enabled, Fake CUDA arrays are passed to some of the
+# @jit-decorated functions. This causes a fallback to object mode and the
+# emission of a warning at present - this is visible in the test suite with
+# cudasim, but also fallback-by-default will not be permitted in future. In
+# order to avoid the warning / future error, we explicitly specify that
+# object mode with loop lifting is acceptable when using the simulator.
+forceobj = looplift = config.ENABLE_CUDASIM
```
Comment (stuartarchibald) [inline numba/cuda/random.py:42]:
```suggestion
_forceobj = _looplift = config.ENABLE_CUDASIM
```
perhaps to imply no-reuse/import from elsewhere?

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the patch, one minor thing to resolve else looks good.

File: numba/cuda/random.py (hunk 1 above)
Comment (gmarkall) [reply numba/cuda/random.py]: Good point - done.

**Comment by stuartarchibald** [comment]
Test failure is due to:
```
Collecting package metadata (current_repodata.json): ...working... done
Solving environment: ...working... done

ChecksumMismatchError: Conda detected a mismatch between the expected content and downloaded content
for url 'https://conda.anaconda.org/numba/label/dev/linux-64/llvmlite-0.37.0dev0-py39hf484d3e_23.tar.bz2'.
  download saved to: /home/vsts/miniconda3/pkgs/llvmlite-0.37.0dev0-py39hf484d3e_23.tar.bz2
  expected md5: 8ea84fa9b7565f84007f73608bf1fde9
  actual md5: d41d8cd98f00b204e9800998ecf8427e



## Package Plan ##

  environment location: /home/vsts/miniconda3/envs/azure_ci

  added / updated specs:
    - llvmlite


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    llvmlite-0.37.0dev0        |  py39hf484d3e_23        19.8 MB  numba/label/dev
    ------------------------------------------------------------
                                           Total:        19.8 MB

The following NEW packages will be INSTALLED:

  llvmlite           numba/label/dev/linux-64::llvmlite-0.37.0dev0-py39hf484d3e_23

```
will retry.

**Comment by gmarkall** [comment]
Thanks! I think the tests are looking good now? (I see no failure)

File: numba/cuda/random.py (hunk 1 above)
Comment (stuartarchibald) [reply numba/cuda/random.py]: thanks

**Review by stuartarchibald — APPROVED** [review]
Thanks for the patch and fixes.

**Comment by stuartarchibald** [comment]
Buildfarm ID: `numba_smoketest_cuda_yaml_37`.

**Comment by stuartarchibald** [comment]
> Buildfarm ID: `numba_smoketest_cuda_yaml_37`.

Passed

**Pull request merged**